### PR TITLE
08_irq: task solution

### PR DESCRIPTION
**Pull request by ASynenko**
gpio_irq implements user button handling with hardware debouncing and and threaded interrupt. 
gpio_poll - software debouncing with timer based polling.

### CODE REVIEW

File: 08_irq/irq/gpio_irq.c [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+		return res;
+	}
+
+	res = (pin->output == true) ? gpio_direction_output(pin->pin_id, 0)
+				: gpio_direction_input(pin->pin_id);
+	if (res < 0)
+		goto failed_gpio_init;
+
+	if (pin->output == false)
+		gpio_set_debounce(pin->pin_id, 20);
```
Comment (ReAlUA) [inline 08_irq/irq/gpio_irq.c]: 20 microseconds is too short time. Perhaps, 2000 is OK in most cases.

File: 08_irq/irq/gpio_irq.c [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+	if (res < 0) {
+		gpio_deinit(&state_led);
+		return res;
+	}
+
+	res = gpio_init(&user_btn);
+	if (res < 0) {
+		gpio_deinit(&busy_led);
+		gpio_deinit(&state_led);
+		return res;
```
Comment (ReAlUA) [inline 08_irq/irq/gpio_irq.c]: In Linux Kernel, `goto` is a preferred way.
Add resource deallocation after the `return 0;`  in a reverse order with a label before each step.

File: 08_irq/poll/gpio_poll.c [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+	pr_info("%s: module starting\n",  __func__);
+
+	timer_setup(&g_timer, &gpio_debounce, 0);
+	res = mod_timer(&g_timer, jiffies + msecs_to_jiffies(20));
+	if (res < 0)
+		return res;
+
+	res = gpio_init(&state_led);
+	if (res < 0)
+		return res;
```
Comment (ReAlUA) [inline 08_irq/poll/gpio_poll.c]: The timer is not deleted here. See comment to the `irq` part.
Also, if this function is interrupted by other drivers, timer callback may be called before GPIO initialization.
Timer start must be the last operation.

File: 08_irq/poll/gpio_poll.c [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+
+static void gpio_deinit(struct gpio_conf *pin)
+{
+	if (pin->irq_id >= 0)
+		free_irq(pin->irq_id, NULL);
+
+	gpio_free(pin->pin_id);
+}
+
+void gpio_debounce(struct timer_list *tm)
```
Comment (ReAlUA) [inline 08_irq/poll/gpio_poll.c:98]: Not implemented: "If the filtered state changes, issue a pulse on LED3 output for 1 ms. Use high resolution timer."

**Review by ReAlUA — CHANGES_REQUESTED** [review]

File: 08_irq/irq/gpio_irq.c (hunk 1 above)
Comment (ASynenko) [reply 08_irq/irq/gpio_irq.c]: Fixed

File: 08_irq/poll/gpio_poll.c (hunk 4 above)
Comment (ASynenko) [reply 08_irq/poll/gpio_poll.c:98]: Changed code to implement.

File: 08_irq/poll/gpio_poll.c (hunk 3 above)
Comment (ASynenko) [reply 08_irq/poll/gpio_poll.c]: Fixed

File: 08_irq/irq/gpio_irq.c (hunk 2 above)
Comment (ASynenko) [reply 08_irq/irq/gpio_irq.c]: Fixed

File: 08_irq/poll/gpio_poll.c [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+	int raw_state = gpio_get_value(user_btn.pin_id);
+
+	if (debounce_state == raw_state)
+		debounce_cnt = 5;
+	else if (!(--debounce_cnt))
+	{
+		debounce_state = raw_state;
+		debounce_cnt = 5;
+		gpio_set_value(busy_led.pin_id, 1);
+		msleep(1);
```
Comment (ReAlUA) [inline 08_irq/poll/gpio_poll.c]: 1. "Use *high resolution timer*"
2. Did you even run this code? Do not sleep in a timer callback. At first, you delay execution of other callbacks on the same jiffies value. At zero, it is an atomic context (timer softirq).

File: 08_irq/poll/gpio_poll.c [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+		goto deinit_sts;
+
+	res = gpio_init(&user_btn);
+	if (res < 0)
+		goto deinit_leds;
+
+	timer_setup(&g_timer, &gpio_debounce, 0);
+	res = mod_timer(&g_timer, jiffies + msecs_to_jiffies(20));
+	if (res == 0)
+		return 0;
```
Comment (ReAlUA) [inline 08_irq/poll/gpio_poll.c]:
```c
if (res)
        goto deinit_btn;

return 0;
```
produces the same code but successful exit point is more clear/
Also, these lines will not be changed if you add some other initialization after them. Just more lines between if/goto and return. Think about future commits/review.

**Review by ReAlUA — CHANGES_REQUESTED** [review]

File: 08_irq/poll/gpio_poll.c [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+{
+	static int debounce_state;
+	static int debounce_cnt = 5;
+
+	int raw_state = gpio_get_value(user_btn.pin_id);
+
+	if (debounce_state == raw_state)
+		debounce_cnt = 5;
+	else if (!(--debounce_cnt))
+	{
```
Comment (ReAlUA) [inline 08_irq/poll/gpio_poll.c]: Run the checkpatch.

File: 08_irq/irq/gpio_irq.c [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+	res = request_threaded_irq(pin->irq_id, pin->irq_handler,
+				pin->irq_thread,
+				IRQF_TRIGGER_RISING | IRQF_TRIGGER_FALLING,
+				"user button hit", NULL);
+
+	if (res >= 0)
+		return res;
+
+failed_gpio_init:
+	pr_err("%s: failed to init gpio pin %d, error %d",
```
Comment (ReAlUA) [inline 08_irq/irq/gpio_irq.c]: '\n' at the end of the format string

File: 08_irq/poll/gpio_poll.c [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+
+	res = request_threaded_irq(pin->irq_id, pin->irq_handler,
+				pin->irq_thread,
+				IRQF_TRIGGER_RISING | IRQF_TRIGGER_FALLING,
+				"user button hit", NULL);
+	if (res >= 0)
+		return res;
+
+failed_gpio_init:
+	pr_err("%s: failed to init gpio pin %d, error %d",
```
Comment (ReAlUA) [inline 08_irq/poll/gpio_poll.c]: '\n' at the end of the format string

File: 08_irq/irq/gpio_irq.c [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+	return IRQ_WAKE_THREAD;
+}
+
+static irqreturn_t btn_thread(int irq, void *prt)
+{
+	gpio_set_value(busy_led.pin_id, 1);
+	msleep(simulate_busy);
+	gpio_set_value(busy_led.pin_id, 0);
+
+	pr_info("%s: button was hit %d time(s)", __func__, hit_counter);
```
Comment (ReAlUA) [inline 08_irq/irq/gpio_irq.c]: '\n' at the end of the format string (check other places too)

**Review by ReAlUA — CHANGES_REQUESTED** [review]

File: 08_irq/irq/gpio_irq.c (hunk 10 above)
Comment (ASynenko) [reply 08_irq/irq/gpio_irq.c]: Fixed

File: 08_irq/poll/gpio_poll.c (hunk 9 above)
Comment (ASynenko) [reply 08_irq/poll/gpio_poll.c]: Fixed

File: 08_irq/irq/gpio_irq.c (hunk 8 above)
Comment (ASynenko) [reply 08_irq/irq/gpio_irq.c]: Fixed

File: 08_irq/poll/gpio_poll.c (hunk 7 above)
Comment (ASynenko) [reply 08_irq/poll/gpio_poll.c]: Done

**Review by ReAlUA — APPROVED** [review]

**Pull request merged**